boards: arm: argonkey: Remove default on choice symbol

ulfalizer opened the pull request:
The choice symbol LSM6DSL_EXT0_LIS2MDL was given a 'default' in
Kboards/arm/96b_argonkey/Kconfig.defconfig, but 'default' has no effect
on choice symbols, triggering a warning from Kconfiglib.

Instead of adding a default to the choice symbol, change the default of
the 'choice' itself by giving it name and adding the default in
Kconfig.defconfig.

This requires that we remove the default on the "base" definition of the
'choice', due to some messiness related to the Zephyr-specific
prefer-later-defaults behavior (see the 'Zephyr-specific Kconfig
behavior for defaults' section in the Board Porting Guide). Choices were
overlooked when that patch was added to the C tools, meaning choices
still prefer earlier defaults. The crux is that Kconfig.defconfig files
are included last.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
CC @avisconti 

See http://docs.zephyrproject.org/porting/board_porting.html#visible-and-invisible-kconfig-symbols for some concepts here. It's not exactly obvious.

erwango commented on boards/arm/96b_argonkey/96b_argonkey_defconfig:
```diff
@@ -47,3 +47,8 @@ CONFIG_CLOCK_STM32_APB2_PRESCALER=1
 CONFIG_UART_CONSOLE=y
+
+# accelerometer and gyroscope sensor
+CONFIG_SENSOR=y
```
According to board porting guide, IPs should configured by default but not configured.

ulfalizer replied:
@erwango 
Yeah, this is mostly just me guessing intent.

What about just removing the `LSM6DSL_EXT0_LIS2MDL` default in `Kconfig.defconfig`? It's a no-op now, and generates a warning. It's really those I'm trying to get rid of, to turn most of them into errors.

That symbol is already the default selection of the choice as well (though it's possible that that could change if someone modified the `choice`).

ulfalizer replied:
People messed up a bit when they added that prefer-later-defaults by the way. They didn't do it for choices. That makes it impossible to override a choice default from a `Kconfig.defconfig` file.

Maybe that could be changed to if it's critical. What I really would want would be to go back to the standard Kconfig behavior for defaults though. :P

ulfalizer replied:
@erwango 
Any comments? Otherwise, I'll change it to just remove the default, which is a no-op.

avisconti replied:
> What about just removing the LSM6DSL_EXT0_LIS2MDL default in Kconfig.defconfig? It's a no-op now, and 
> generates a warning. It's really those I'm trying to get rid of, to turn most of them into errors.

ok, I see.


> That symbol is already the default selection of the choice as well (though it's possible that that could change if 
> someone modified the choice).

Yes, it could change in the future. To be honest I would like to enforce it here in the board section. How can we do that in a clean way?

@erwango

ulfalizer replied:
@erwango @avisconti 
See [this section](http://docs.zephyrproject.org/porting/board_porting.html#zephyr-specific-kconfig-behavior-for-defaults) in the Board Porting Guide for an overview of the prefer-later-defaults behavior.

That patch was in the C tools too, and [got inherited](https://github.com/zephyrproject-rtos/Kconfiglib/commit/600033bed8dd0ae1068674b379c737c8d0982de4) by Kconfiglib.

The problem is that `choice`s were overlooked and still prefer earlier defaults, meaning it's impossible to override a `default` from the "base" definition of a `choice` from a `Kconfig.defconfig` file (the crux is that the `Kconfig.defconfig` files are included last).

Maybe prefer-later-defaults could be implemented for choices as well, though you'd need to make sure that you don't break any existing choices that depend on the inconsistency. Bit messy...

I pushed a new version that just gets rid of the `default` before I saw your comments by the way. My current focus is on getting rid of certain classes of errors from the build, to turn them into errors.

ulfalizer replied:
@avisconti 
If the base definition of the choice didn't have a default, it would be possible to define the choice twice though to specify the default, like this:

**drivers/sensor/lsm6dsl/Kconfig:**
```
choice EXTERNAL_SENSOR_0
	prompt "External sensor 0"
	help
	  Choose the external sensor 0 connected to LSM6DS3.

config LSM6DSL_EXT0_LIS2MDL
	bool "LIS2MDL"

config LSM6DSL_EXT0_LPS22HB
	bool "LPS22HB"

endchoice
```

**boards/arm/96b_argonkey/Kconfig.defconfig:**
```
choice EXTERNAL_SENSOR_0
	default LSM6DSL_EXT0_LIS2MDL
endchoice
```

Would that work in this case?

Maybe it could be cleaned up separately. I'm in batch warning mode atm.

avisconti replied:
@ulfalizer 
Using a named choice?
Well, actually I was thinking the same... Yes, maybe it works.
What do you think @erwango?
I can try the commit out if you push it.

ulfalizer replied:
@avisconti 
I pushed a new version that uses a named choice (and gets rid of the `default` on the base definition of the `choice`).

erwango replied:
@ulfalizer , @avisconti , if named choice is an acceptable solution, then let's use it

ulfalizer replied:
@erwango 
It's the only way to change the default of a choice at least. It's unfortunate that you have to remove the default on the base definition though, due to that later-defaults oversight.

erwango commented:
@ulfalizer , I agree we need to make some clean up in Kconfig symbols and it is good to have some rules. 
Though, I think we should first think if this is right for the symbol to be user assignable then check if prompt is set accordingly and then put it in the right file.
I think that for a good part of the Kconfig symbols, the "user assignable" property has not been verified before introducing the prompt, so we should verify this first and then clean up _config/.defconfig files.

So good question here would be:
Should the option to choose sensor connected to EXT0 pin be selectable in menuconfig ?

ulfalizer committed:
boards: arm: argonkey: Properly set choice default

The choice symbol LSM6DSL_EXT0_LIS2MDL was given a 'default' in
Kboards/arm/96b_argonkey/Kconfig.defconfig, but 'default' has no effect
on choice symbols, triggering a warning from Kconfiglib.

Instead of adding a default to the choice symbol, change the default of
the 'choice' itself by giving it name and adding the default in
Kconfig.defconfig.

This requires that we remove the default on the "base" definition of the
'choice', due to some messiness related to the Zephyr-specific
prefer-later-defaults behavior (see the 'Zephyr-specific Kconfig
behavior for defaults' section in the Board Porting Guide). Choices were
overlooked when that patch was added to the C tools, meaning choices
still prefer earlier defaults. The crux is that Kconfig.defconfig files
are included last.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

avisconti commented:
@ulfalizer @erwango 
I cherry-picked the 31eb582 and it works fine as before!
So, I think we could proceed in this way!

Thx!

ulfalizer commented:
Great!

@erwango 
Are you fine with this too?

ulfalizer commented:
@avisconti 
Yep, fixes the warning re. the `default` on the choice symbol.

By the way, an easy way to check is to run `make kconfig` in `doc/`. That'll parse all the Kconfigs. It's used when generating the symbol reference.
